### PR TITLE
add distributed optimization implementation

### PR DESCRIPTION
**Pull request by WenboRen**
This branch is for adding data parallelism distributed optimization algorithms to DLRM on top of the version with model parallelism implemented. The algorithms implemented include local SGD, post-local SGD, hierarchical local SGD, noise injection, and slow momentum. For these algorithms, we have done numerous experiments on Criteo Kaggle and Criteo Terabyte.

### CODE REVIEW

**Comment by mnaumovfb** [comment]
Looks good. I only have a few minor comments/suggestions:

1. Can you clarify the reason behind the change in md_flag? 
https://github.com/facebookresearch/dlrm/pull/107/files#diff-e17ed39abb7ba54648f2628e4e955a56R160

2.  Can you clarify the print frequency change with should_print flag? 
https://github.com/facebookresearch/dlrm/pull/107/files#diff-e17ed39abb7ba54648f2628e4e955a56R1221 
https://github.com/facebookresearch/dlrm/pull/107/files#diff-e17ed39abb7ba54648f2628e4e955a56R1225
I would prefer to keep the original behavior unless there are bugs.

3. Also, let's use the new ONNX code for saving the model 
https://github.com/facebookresearch/dlrm/pull/107/files#diff-e17ed39abb7ba54648f2628e4e955a56R1138
https://github.com/facebookresearch/dlrm/pull/107/files#diff-e17ed39abb7ba54648f2628e4e955a56R1468

File: dlrm_s_pytorch.py [hunk 1]
```diff
@@ -91,45 +97,8 @@
 # import torch.nn.functional as Functional
 # from torch.nn.parameter import Parameter
 
-from torch.optim.lr_scheduler import _LRScheduler
-
 exc = getattr(builtins, "IOError", "FileNotFoundError")
 
-class LRPolicyScheduler(_LRScheduler):
```
Comment (ddkalamk) [inline dlrm_s_pytorch.py:98]: Can we please keep these LR scheduler related changes?

**Comment by WenboRen** [comment]
> Looks good. I only have a few minor comments/suggestions:
> 
> 1. Can you clarify the reason behind the change in md_flag?
>    https://github.com/facebookresearch/dlrm/pull/107/files#diff-e17ed39abb7ba54648f2628e4e955a56R160

I did experiments on Kaggle and Terabyte and found that this change is not necessary. This change has been reverted.

> 2. Can you clarify the print frequency change with should_print flag?
>    https://github.com/facebookresearch/dlrm/pull/107/files#diff-e17ed39abb7ba54648f2628e4e955a56R1221
>    https://github.com/facebookresearch/dlrm/pull/107/files#diff-e17ed39abb7ba54648f2628e4e955a56R1225
>    I would prefer to keep the original behavior unless there are bugs.

Inside the function distributed_forward() (line 372), if batch_size % ext_dist.my_size != 0, then the program will exit with errors. For some values of world_size (ext_dist.my_size) and batch-size, the size of the last batch may not be divided by ext_dist.my_size, and thus, we added a logic to skip the last batch before distributed_forward() is called. Due to this, the testing and printing in the last batch will also be skipped. Thus, we moved the final printing and testing one batch ahead. Some comments were added to explain this. 

> 3. Also, let's use the new ONNX code for saving the model
>    https://github.com/facebookresearch/dlrm/pull/107/files#diff-e17ed39abb7ba54648f2628e4e955a56R1138
>    https://github.com/facebookresearch/dlrm/pull/107/files#diff-e17ed39abb7ba54648f2628e4e955a56R1468

Sure. I have changed to the new ONNX code.

File: dlrm_s_pytorch.py (hunk 1 above)
Comment (WenboRen) [reply dlrm_s_pytorch.py:98]: Thanks for the suggestion. I have added back the LRScheduler in the newest commit.
